doc/crdt-details.md

```diff
@@ -0,0 +1,1054 @@
+    Engine:   text="ab" tombstones="x123" deletes_from_union={1,3,4,5}
+    Concepts: union="axb123"
+```
+
+## Operations
+
+Now that you know what we have to work with, let's go over the operations that `Engine` supports. Each of these operations relies on a bunch of different shared helpers, as we go from the simplest operations to the most complex, we'll gradually build up the set of helpers we use.
+
+After describing how each operation or helper works there'll often be a code block with the actual function, it isn't necessary to understand the code and it may have complications that aren't mentioned, so feel free to skip them, they're there if you want to confirm or enhance your understanding.
+
+### Subset helpers
+
+`Subset` has a number of operations that produce new `Subset`s, these form the core of most work the CRDT operations do. It's better to explain them together near the start since they're used everywhere:
+
+#### Subset::union
+
+Takes two `Subset`s and produces a new `Subset` of the same string where each character has the sum of the counts it has in each input. When treating a `Subset` as a normal set, this is just the union.
+
+#### Subset::transform_expand
+
+`Subset::transform_expand` takes a `Subset` and another "transform" `Subset` and transforms the first `Subset` through the coordinate transform represented by the "transform". Now what does this mean:
+
+`Revision`s are never modified even when the union string expands with newly inserted characters. We can deal with this by treating `Subset`s of inserted characters as coordinate transforms. The only difference to the union string is the inserted characters, so if we can modify the coordinates of a `Subset` from being based on one union string to another, we can work with edits from multiple `Revision`s together.
+
+We can do this by "expanding" the indices in a `Subset` after each insert by the size of that insert, where the inserted characters are the "transform". Conceptually if a `Subset` represents the set of characters in a string that were inserted by an edit, then it can be used as a transform from the coordinate space before that edit to after that edit by mapping a `Subset` of the string before the insertion onto the 0-count regions of the transform `Subset`.
+
+The actual procedure works by iterating over the segments of the transform:
+
+- When it encounters a 0-count segment it keeps putting (potentially partial) segments from `self` into the output until it fills its size.
+- When it encounters a non-0-count transform segment it outputs a 0-count segment of the same size.
+
+The `SubsetBuilder` used for the output automatically merges consecutive segments that have the same count. See the diagram below for how this plays out:
+
+![transform_expand workings](img/trans-expand-1.png)
+
+One example of how this can be used is to find the characters that were inserted by a past `Revision` in the coordinates of the current union string instead of the past one:
```

Comment:
One example of how this can be used is to find the characters that were inserted by a past `Revision` in the coordinates of ***another `Revision`'s union string:*** ~the current union string instead of the past one:~